autoarray/inversion/inversion/inversion_util.py

```diff
@@ -289,6 +294,88 @@ def reconstruction_from(
+    )
+
+    return reconstruction
+
+
+def reconstruction_positive_only_from(
+    data_vector: np.ndarray,
+    curvature_reg_matrix: np.ndarray,
+    settings: SettingsInversion = SettingsInversion(),
+):
+    """
+    Solve the linear system [F + reg_coeff*H] S = D -> S = [F + reg_coeff*H]^-1 D given by equation (12)
+    of https://arxiv.org/pdf/astro-ph/0302587.pdf
+
+    S is the vector of reconstructed inversion values.
+
+    This reconstruction uses a linear algebra solver that allows only positives values in the solution.
+    By not allowing negative values, the solver is slower than methods which allow negative values, but there are
+    many inference problems where negative values are nonphysical or undesirable and removing them improves the solution.
+
+    This function checks that the solution does not give a linear algebra error (e.g. because the input matrix is
+    not positive-definitive) and that it avoids solutions where all reconstructed values go to the same value. If these
+    occur it raises an exception.
+
+    Parameters
+    ----------
+    data_vector
+        The `data_vector` D which is solved for.
+    curvature_reg_matrix_cholesky
+        The sum of the curvature and regularization matrices.
+    settings
+        Controls the settings of the inversion, for this function where the solution is checked to not be all
+        the same values.
+
+    Returns
+    -------
+    curvature_reg_matrix
+        The curvature_matrix plus regularization matrix, overwriting the curvature_matrix in memory.
+    """
+
+    try:
+
+        reconstruction = nnls(
+            curvature_reg_matrix,
+            (data_vector).T,
+            maxiter=settings.positive_only_maxiter,
+        )[0]
+
+    except (RuntimeError, np.linalg.LinAlgError):
+        raise exc.InversionException()
+
+    reconstruction_check_solution(
+        data_vector=data_vector, reconstruction=reconstruction, settings=settings
+    )
+
+    return reconstruction
+
+
+def reconstruction_check_solution(
+    data_vector: np.ndarray,
+    reconstruction: np.ndarray,
+    settings: SettingsInversion = SettingsInversion(),
+):
+    """
+    Check that the solution after reconstructing an inversion is not one where all reconstructed values go to the same
+    value.
+
+    These solutions occur when the linear system is numerically unstable, and can lead to high `log_evidence` values
+    due to numerically issues not currently fully understood.
+
+    Parameters
+    ----------
+    data_vector
+        The `data_vector` D which is solved for.
+    curvature_reg_matrix
+        The the sum of the curvature and regularization matrices.
+    settings
+        Controls the settings of the inversion, for this function where the solution is checked to not be all
+        the same values.
+    """
     if settings.check_solution and len(reconstruction) > 1:
         if np.isclose(a=reconstruction[0], b=reconstruction[1], atol=1e-6).all():

```

Comment:
If reconstructure can have length greater than two do you instead want to check allclose?

```python
np.allclose(reconstruction[0], reconstruction[1:])
```

This returns True if len(reconstruction) == 1 so may allow you to remove the length check